src/intrinsics/ecma262/ArrayPrototype.js

```diff
@@ -700,6 +712,39 @@ export default function(realm: Realm, obj: ObjectValue): void {
     // 1. Let O be ? ToObject(this value).
     let O = To.ToObject(realm, context);
 
+    if (
+      realm.isInPureScope() &&
+      O instanceof AbstractObjectValue &&
+      (O.kind === "Array.from(A,B,C)" || O.kind === "(A).map(B,C)" || O.kind === "Object.keys(A)")
+    ) {
+      let args = [context, callbackfn];
+      let safeToCreateTemporal = true;
+
+      // if mapfn or thisArg exist and are not statically known,
+      // we cannot guarantee that something will be not be havoced
+      // so rather, we skip this logic and continue with the default
+      // behaviur of Array.prototype.map
+      if (!(callbackfn instanceof ECMAScriptSourceFunctionValue || callbackfn instanceof BoundFunctionValue)) {
+        safeToCreateTemporal = false;
+        if (thisArg) {
+          if (!(thisArg instanceof AbstractValue)) {
+            safeToCreateTemporal = false;
+          }
+          args.push(thisArg);
+        }
+      }
+
+      if (safeToCreateTemporal) {
+        return AbstractValue.createAbstractTemporalArray(
```

Comment:
Whenever we generate temporal values like these that might throw, we have to wrap it in `realm.evaluateWithPossibleThrowCompletion`. We've been sloppy with that but we should be more careful about that.